bin/cheribsdtest/cheribsdtest_util_vm.c

```diff
@@ -0,0 +1,97 @@
+/*-
+ * Copyright (c) 2021, 2022 SRI International
+ * All rights reserved.
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory under DARPA/AFRL contract (FA8750-10-C-0237)
+ * ("CTSRD"), as part of the DARPA CRASH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+
+#include <sys/param.h>
+#include <sys/sysctl.h>
+#include <sys/user.h>
+
+#include <cheri/cheric.h>
+
+#include <libprocstat.h>
+#include <unistd.h>
+
+#include "cheribsdtest.h"
+
+/*
+ * Find a region of address space unoccupied by any memory mappings.
+ *
+ * Caveats:
+ * - This is not concurency safe and the found region could be disrupted
+ *   by any non-MAP_FIXED mmap() call.  This includes anything that
+ *   allocates heap memory.
+ * - The region between PAGE_SIZE and the first mapping is not currently
```

Comment:
Nor is between the end of the last mapping and the end of the user-mappable address space.

I guess there's also the caveat that this wouldn't work on architectures with a non-contiguous user address space (Itanium comes to mind), but probably rather out of scope...

Reply:
Currently the last page is always the shared page. Without that we might need to add some handling of the top of VA at least on coexecve were the first process's initial stack will eventually go away.